File: src/rosjava_pkg/odometry/src/main/java/odometry/Odometry.java
package odometry;

import org.ros.message.MessageListener;
import rss_msgs.EncoderMsg;
import rss_msgs.OdometryMsg;
import org.ros.namespace.GraphName;
import org.ros.node.Node;
import org.ros.node.NodeMain;
import org.ros.node.ConnectedNode;
import org.ros.node.topic.Publisher;
import org.ros.node.topic.Subscriber;
public class Odometry implements NodeMain {

	private int[] prev_ticks; // (left wheel, right wheel, make int[]
	private static final double WHEEL_RADIUS_IN_M = 0.0625;
	private static final double ENCODER_RESOLUTION = 2000;
	private static final double GEAR_RATIO = 65.5;
	private static final double TICKS_PER_REVOLUTION=ENCODER_RESOLUTION * GEAR_RATIO;
	private static final double WHEEL_METERS_PER_TICK = WHEEL_RADIUS_IN_M * Math.PI * 2/(TICKS_PER_REVOLUTION);
	private static final double WHEELBASE = .428;
	/**
	 * [x][y][theta]
	 */
	private double[] pose = {0, 0, 0};
	private boolean reset = false;
	private Publisher<rss_msgs.OdometryMsg> pub;
	private Publisher<std_msgs.String> statePub;
	private Subscriber<rss_msgs.EncoderMsg> encoderSub;
	private Subscriber<rss_msgs.OdometryMsg> resetOdometrySub;
        private rss_msgs.OdometryMsg msg = pub.newMessage();

	public void update(int[] new_ticks) {
		if ((prev_ticks == null && (new_ticks[0]==0 && new_ticks[1]==0)) || reset) {
			prev_ticks = new_ticks;
			reset = false;
		}
		else if (prev_ticks == null) {
			prev_ticks = new int[2];
		}
		int[] ticks = new int[2];
		for (int i = 0; i < 2; i ++){
			ticks[i] = new_ticks[i] - prev_ticks[i];
		}
		if (ticks[0] == 0 && ticks[1] == 0) {
			pub.publish(msg);
			return; // we haven't moved at all
		}

		double s_left = (ticks[0])*WHEEL_METERS_PER_TICK;
		double s_right = (ticks[1])*WHEEL_METERS_PER_TICK;
		double theta = (s_left - s_right)/WHEELBASE;
		msg.setTheta( msg.getTheta()-theta);
		if(msg.getTheta() < 0){
			msg.setTheta(msg.getTheta()+2*Math.PI);
		} else if (msg.getTheta() > 2*Math.PI){
			msg.setTheta(-2*Math.PI+(msg.getTheta()));
		}
		msg.setX((s_left+s_right)*Math.cos(msg.getTheta())/2.0+(msg.getX()));
		msg.setY((s_left+s_right)*Math.sin(msg.getTheta())/2.0+(msg.getY()));

		prev_ticks[0] = new_ticks[0];
		prev_ticks[1] = new_ticks[1];
		pub.publish(msg);
	}

	@Override
	public void onShutdown(Node node) {
		// TODO Auto-generated method stub

	}

	@Override
	public void onStart(ConnectedNode node) {
		pub = node.newPublisher("/rss/odometry", "rss_msgs/OdometryMsg");
		encoderSub = node.newSubscriber("/rss/Encoder", "rss_msgs/EncoderMsg");
		encoderSub.addMessageListener(new EcoderListener(this));

		resetOdometrySub = node.newSubscriber("/rss/odometry_update", "rss_msgs/OdometryMsg");
		resetOdometrySub.addMessageListener(new MessageListener<rss_msgs.OdometryMsg>() {
			@Override
			public void onNewMessage(rss_msgs.OdometryMsg message) {
				if (message.getX() == 0 && message.getY() == 0 && message.getTheta() == 0) {
					prev_ticks = null;
					msg.setX(message.getX()); msg.setY(message.getY()); msg.setTheta(message.getTheta());
					reset = true;
				}
			}
		});
	}



	@Override
	public void onShutdownComplete(Node node) {
	}

	@Override
	public GraphName getDefaultNodeName() {
		return GraphName.of("rss/odometry");
	}
        @Override
	public void onError(Node node, Throwable thrown){
        }
}

File: src/odometry/src/odometry/odometry.py
#!/usr/bin/env python
import rospy
import math
import time
from gc_msgs.msg import PoseMsg
from gc_msgs.msg import EncoderMsg
from gc_msgs.msg import StateMsg
import location 
import pose

#
# This node recieves encoder messages, calculates the robot's current position, and sends out the robot's current position.
# It also keeps track of the robot's uncertainty in position, which it updates upon recieving an uncertainty message. The uncertainty is sent out in the position message. The position includes (x position, y position, angle)
#


class Odometry(object):
    
    # params: none
    # returns: none
    # assigns constants, create publishers, init node
    def __init__(self):
        # TODO
        # initializing node
        rospy.init_node('odometry')
        #rospy.onShutdown(self.onShutdown)

        # initial values
        self.START_POSE = pose.Pose(0.0, 0.0, 0.0)

        # global variables
        self.currentDist = None        # left, right values of ticks from last update
        self.currentPose = self.START_POSE    # xPosition of robot origin, yPosiiton of robot origin
                                    #   angle of robot axis of symmetry with respect to x axis.
                                    #   All pose fields are floats

        #constants
        #TODO: get right values
        self.WHEELBASE = .372       # in m
                               # TODO: measure
        self.LEFT = 0              # for indexing into encoder lists
        self.RIGHT = 1             # for indexing into encoder lists
    

        #publishers and subscribers
        #TODO: uncertainty publisher and subscriber

        self.posePub = rospy.Publisher('sensor/currentPose', PoseMsg);
        self.encoderSub = rospy.Subscriber('sensor/Encoder', EncoderMsg, self.handleEncoderMsg);
        self.stateSub = rospy.Subscriber('command/State', StateMsg, self.handleStateMsg)


        return

####################
# helper functions #
####################

    # param: none
    # return: none
    # on shutdown behavior
    def onShutdown(self):
        return

    # params: deltaTicks: int[] [new ticks on left wheel, new ticks on right wheel]
    # returns: void
    # updates currentPose given new Ticks, sends message 
    def updateCurrentPose(self, deltaDist):

        # calculating distance travelled by each wheel
        leftWheelDist = deltaDist[self.LEFT]
        rightWheelDist = deltaDist[self.RIGHT]
    
        # calculating and changing theta
        deltaTheta = (leftWheelDist - rightWheelDist)/self.WHEELBASE         # using small angle approximation
        newAngle = (self.currentPose.getAngle() - deltaTheta)%(2*math.pi)     # calculating new angle, wrapping aroun 2pi

        self.currentPose.setAngle(newAngle)

        # calculating and changing x, y position
        deltaX = (leftWheelDist + rightWheelDist)*math.cos(self.currentPose.getAngle())/2.0
        deltaY = (leftWheelDist + rightWheelDist)*math.sin(self.currentPose.getAngle())/2.0
        self.currentPose.setX(self.currentPose.getX() + deltaX)
        self.currentPose.setY(self.currentPose.getY() + deltaY)

        # sending updated currentPose
        self.reportCurrentPose()
        
        return


################################
# recieving, sending messages  #
################################

    # param: State Msg
    # returns: none
    # resets currentPose to StartPose if get 'init' msg
    def handleStateMsg(self, msg):
        if msg.state == 'init':
            self.currentPose = self.START_POSE
            self.currentDist = None
            print "I reset!"
        return


    # params: EncoderMsg msg
    # returns: none
    # updates currentPosition based on the message; sends out updated currentPosition value (xPos, yPos, angle)
    # assumes encoder message fixes signs so both wheels have posive ticks when the robot is moving forward
    def handleEncoderMsg(self, msg):
        
        if self.currentDist == None:
            self.currentDist = [msg.lWheelDist, msg.rWheelDist]
        
        # calculating how much the wheels have moved in the past time step
        newDist = [msg.lWheelDist, msg.rWheelDist] # current tick positions of the wheels
        deltaDist = [newDist[self.LEFT] - self.currentDist[self.LEFT], newDist[self.RIGHT]-self.currentDist[self.RIGHT]]


        # updating the current position and sending
        self.updateCurrentPose(deltaDist)
        self.currentDist = newDist

        return

    # params: uncertaintyMsg
    # returns: none
    # updates uncertainty based on the message; sends out updates currentPosition value
    def handleUncertaintyMsg(self, msg):
        # TODO
        pass

    # params: none
    # returns none
    # sends message containing currentPose
    def reportCurrentPose(self):
        # forming message from currentPose
        msg = PoseMsg()
        msg.xPosition = self.currentPose.getX()
        msg.yPosition = self.currentPose.getY()
        msg.angle = self.currentPose.getAngle()
        
        self.posePub.publish(msg)
        
        return

#########
## main #
#########
if __name__ == '__main__':


    try:
        odometry = Odometry()
        rospy.spin()                    # keep python from exiting until this node is stopped 
                                        # wait for odometry interrupts
    except rospy.ROSInterruptException: pass
